numpy/f2py/cfuncs.py

```diff
@@ -548,24 +548,31 @@ def errmess(s: str) -> None:
 #error You need to install NumPy version 0.13 or higher. See https://scipy.org/install.html
 #endif
 """
+
+# Defining the correct value to indicate thread-local storage in C without
+# running a compile-time check (which we have no control over in generated
+# code used outside of NumPy) is hard. Therefore we support overriding this
+# via an external define - the f2py-using package can them use the same
+# compile-time checks as we use for `NPY_TLS` when building NumPy.
+#
+# __STDC_NO_THREADS__ should not be coupled to the availability of _Thread_local.
+# In case we get a bug report, guard it with __STDC_NO_THREADS__ after all.
+#
+# `thread_local` has become a keyword in C23, but don't try to use that yet
+# (too new, doing so while C23 support is preliminary will likely cause more
+#  problems than it solves).
+#
+# Note: do not try to use `threads.h`, its availability is very low
+# *and* threads.h isn't actually used where `F2PY_THREAD_LOCAL_DECL` is
+# in the generated code. See gh-27718 for more details.
 cppmacros["F2PY_THREAD_LOCAL_DECL"] = """
 #ifndef F2PY_THREAD_LOCAL_DECL
 #if defined(_MSC_VER)
 #define F2PY_THREAD_LOCAL_DECL __declspec(thread)
 #elif defined(NPY_OS_MINGW)
 #define F2PY_THREAD_LOCAL_DECL __thread
-#elif defined(__STDC_VERSION__) \\
-      && (__STDC_VERSION__ >= 201112L) \\
-      && !defined(__STDC_NO_THREADS__) \\
-      && (!defined(__GLIBC__) || __GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ > 12)) \\
-      && !defined(NPY_OS_OPENBSD) && !defined(NPY_OS_HAIKU)
-/* __STDC_NO_THREADS__ was first defined in a maintenance release of glibc 2.12,
-   see https://lists.gnu.org/archive/html/commit-hurd/2012-07/msg00180.html,
-   so `!defined(__STDC_NO_THREADS__)` may give false positive for the existence
-   of `threads.h` when using an older release of glibc 2.12
-   See gh-19437 for details on OpenBSD */
-#include <threads.h>
-#define F2PY_THREAD_LOCAL_DECL thread_local
+#elif defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 201112L)
+#define F2PY_THREAD_LOCAL_DECL _Thread_local
```

Comment:
I did confirm that this feature restores my ability to build latest SciPy `main` locally with GNU toolchain on MacOS 14.7 when compared with the latest stable release of NumPy, i.e., via `CC=gcc-13 CXX=g++-13 FC=gfortran-13 python dev.py build -j 16`.

I'm not sure I understand all the details, but it sounds like other projects can customize behavior via the typical build system `try_compile` or "build small program, make define decision based on actual compile outcome." approach.

Perhaps unrelated/another matter, but just in case, I did notice that the actual testsuite for SciPy on ARM mac in this situation with OpenBLAS + GNU toolchain does hard crash at test time:

<details>

```
Fatal Python error: Segmentation fault

Thread 0x000000016c663000 (most recent call first):
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/execnet/gateway_base.py", line 534 in read
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/execnet/gateway_base.py", line 567 in from_io
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/execnet/gateway_base.py", line 1160 in _thread_receiver
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/execnet/gateway_base.py", line 341 in run
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/execnet/gateway_base.py", line 411 in _perform_spawn

Current thread 0x00000001f8a20f40 (most recent call first):
  File "<frozen importlib._bootstrap>", line 488 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 1289 in create_module
  File "<frozen importlib._bootstrap>", line 813 in module_from_spec
  File "<frozen importlib._bootstrap>", line 921 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1331 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1360 in _find_and_load
  File "<frozen importlib._bootstrap>", line 488 in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1415 in _handle_fromlist
  File "/Users/treddy/github_projects/scipy/build-install/lib/python3.12/site-packages/scipy/spatial/distance.py", line 118 in <module>
  File "<frozen importlib._bootstrap>", line 488 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 995 in exec_module
  File "<frozen importlib._bootstrap>", line 935 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1331 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1360 in _find_and_load
  File "/Users/treddy/github_projects/scipy/build-install/lib/python3.12/site-packages/scipy/spatial/_geometric_slerp.py", line 9 in <module>
  File "<frozen importlib._bootstrap>", line 488 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 995 in exec_module
  File "<frozen importlib._bootstrap>", line 935 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1331 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1360 in _find_and_load
  File "/Users/treddy/github_projects/scipy/build-install/lib/python3.12/site-packages/scipy/spatial/__init__.py", line 116 in <module>
  File "<frozen importlib._bootstrap>", line 488 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 995 in exec_module
  File "<frozen importlib._bootstrap>", line 935 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1331 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1360 in _find_and_load
  File "<frozen importlib._bootstrap>", line 488 in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1310 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1360 in _find_and_load
  File "/Users/treddy/github_projects/scipy/build-install/lib/python3.12/site-packages/scipy/cluster/vq.py", line 74 in <module>
  File "<frozen importlib._bootstrap>", line 488 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 995 in exec_module
  File "<frozen importlib._bootstrap>", line 935 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1331 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1360 in _find_and_load
  File "<frozen importlib._bootstrap>", line 488 in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1415 in _handle_fromlist
  File "/Users/treddy/github_projects/scipy/build-install/lib/python3.12/site-packages/scipy/cluster/__init__.py", line 27 in <module>
  File "<frozen importlib._bootstrap>", line 488 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 995 in exec_module
  File "<frozen importlib._bootstrap>", line 935 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1331 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1360 in _find_and_load
  File "<frozen importlib._bootstrap>", line 1387 in _gcd_import
  File "<frozen importlib._bootstrap>", line 488 in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1310 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1360 in _find_and_load
  File "<frozen importlib._bootstrap>", line 1387 in _gcd_import
  File "<frozen importlib._bootstrap>", line 488 in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1310 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1360 in _find_and_load
  File "<frozen importlib._bootstrap>", line 1387 in _gcd_import
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/__init__.py", line 90 in import_module
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/_pytest/pathlib.py", line 582 in import_path
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/_pytest/python.py", line 493 in importtestmodule
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/_pytest/python.py", line 546 in _getobj
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/_pytest/python.py", line 284 in obj
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/_pytest/python.py", line 562 in _register_setup_module_fixture
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/_pytest/python.py", line 549 in collect
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/_pytest/runner.py", line 389 in collect
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/_pytest/runner.py", line 341 in from_call
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/_pytest/runner.py", line 391 in pytest_make_collect_report
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/_pytest/runner.py", line 567 in collect_one_node
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/_pytest/main.py", line 835 in _collect_one_node
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/_pytest/main.py", line 970 in genitems
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/_pytest/main.py", line 975 in genitems
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/_pytest/main.py", line 975 in genitems
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/_pytest/main.py", line 975 in genitems
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/_pytest/main.py", line 809 in perform_collect
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/_pytest/main.py", line 347 in pytest_collection
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/_pytest/main.py", line 336 in _main
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/_pytest/main.py", line 283 in wrap_session
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/_pytest/main.py", line 330 in pytest_cmdline_main
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/xdist/remote.py", line 393 in <module>
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/execnet/gateway_base.py", line 1291 in executetask
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/execnet/gateway_base.py", line 341 in run
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/execnet/gateway_base.py", line 411 in _perform_spawn
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/execnet/gateway_base.py", line 389 in integrate_as_primary_thread
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/execnet/gateway_base.py", line 1273 in serve
  File "/Users/treddy/python_venvs/py_312_scipy_dev/lib/python3.12/site-packages/execnet/gateway_base.py", line 1806 in serve
  File "<string>", line 8 in <module>
  File "<string>", line 1 in <module>

Extension modules: numpy._core._multiarray_umath, numpy.linalg._umath_linalg, cython.cimports.libc.math, scipy._lib._ccallback_c, scipy._lib._fpumode, numpy.random._common, numpy.random.bit_generator, numpy.random._bounded_integers, numpy.random._mt19937, numpy.random.mtrand, numpy.random._philox, numpy.random._pcg64, numpy.random._sfc64, numpy.random._generator, charset_normalizer.md, scipy._lib._test_ccallback, scipy.sparse._sparsetools, _csparsetools, scipy.sparse._csparsetools, scipy.linalg._fblas, scipy.linalg._flapack, scipy.linalg.cython_lapack, scipy.linalg._cythonized_array_utils, scipy.linalg._solve_toeplitz, scipy.linalg._decomp_lu_cython, scipy.linalg._matfuncs_sqrtm_triu, scipy.linalg._matfuncs_expm, scipy.linalg._linalg_pythran, scipy.linalg.cython_blas, scipy.linalg._decomp_update, scipy.sparse.linalg._dsolve._superlu, scipy.sparse.linalg._eigen.arpack._arpack, scipy.sparse.linalg._propack._spropack, scipy.sparse.linalg._propack._dpropack, scipy.sparse.linalg._propack._cpropack, scipy.sparse.linalg._propack._zpropack, scipy.sparse.csgraph._tools, scipy.sparse.csgraph._shortest_path, scipy.sparse.csgraph._traversal, scipy.sparse.csgraph._min_spanning_tree, scipy.sparse.csgraph._flow, scipy.sparse.csgraph._matching, scipy.sparse.csgraph._reordering, scipy.spatial._ckdtree, scipy._lib.messagestream, scipy.spatial._qhull, scipy.spatial._voronoi (total: 47)

```

</details>

that seems likely to be a separate matter though

Reply:
Thanks! Is the crash the same with your old workaround as with this branch? If so then I agree it's separate.